### DISCUSSION TITLE
python3问题

### DISCUSSION THREAD

**Discussion by dongfangrui** [macvim-dev/macvim#1201] [Q&A]
echo has（‘python3’） 显示0，导致很多插件报错

**Comment by eirnym** [comment]
May I ask to write a question in English? This way a question has higher chance to be answered

**Reply by eirnym** [reply]
After translating your question by a translate service, I additionally to writing in English suggest to:

1) check if python 3 is installed in paths MacVim expects to find it (e.g. Homebrew, MacPorts or an official installer from python.org)
2) make sure that MacVim and python are built for the same architecture (x32, x64 or aarch64)

If these suggestions are not working, write a bug report, filling whole form and test different cases mentioned there.

**Comment by dongfangrui** [comment]
No matter if you call "brew install macvim" or download the dmg file here and install it directly, the result is the same. When you open macvim and enter echo has('python3') on the command line, the displayed result is 0. Obviously, python3 is not supported. If you enter echo has('python'), the display result is 1, which obviously means that python2 is supported.

**Comment by dongfangrui** [comment]
In addition, if you do not use macvim, but use the terminal to open vim, and then enter "echo has(python3)", the displayed result is 1, indicating that the vim of the macos system itself supports python3, and of course also supports pytho2.

**Reply by eirnym** [reply]
**Environment (please complete the following information):**
 - Vim version [e.g. 8.1.1234] (Or paste the result of `vim --version`.)
 - OS: [e.g, macOS 10.14, intel x64]
 - Terminal: [e.g. GNOME Terminal, mintty, iTerm2, tmux, GNU screen] (Use GUI if you use the GUI.)
- MacVim architecture [e.g. Intel x64, M1]
- How MacVim was installed [Homebrew, MacPorts, GitHub releases, self built]
- Python versions installed, their architecture
- Python installation source 
- Homebrew, MacPorts architecture if installed

**Comment by dongfangrui** [comment]
Vim version                         ：VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jul 08 2021 16:05:00)
OS                                       ：macOS version - big sur 11.4 (20F71),x86_64 包含补丁: 1-3125
Terminal                              ：macOS itself terminal                
macvim architecture          ：Intel x64
MacVim installed                ：Homebrew and Github-dmg tried all
versions architecture         ：3.9.6，GUI
Python installation source ：https://www.python.org/downloads/。brew install python3
Homebrew architecture     ：3.2.1,Intel

**Comment by dongfangrui** [comment]
localhost:~ xx$ vim --version
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jul 08 2021 16:05:00)
macOS version - x86_64
包含补丁: 1-3125
编译者 Homebrew
巨型版本 无图形界面。  可使用(+)与不可使用(-)的功能:
+acl               +file_in_path      +mouse_urxvt       -tag_any_white
+arabic            +find_in_path      +mouse_xterm       -tcl
+autocmd           +float             +multi_byte        +termguicolors
+autochdir         +folding           +multi_lang        +terminal
-autoservername    -footer            -mzscheme          +terminfo
-balloon_eval      +fork()            +netbeans_intg     +termresponse
+balloon_eval_term +gettext           +num64             +textobjects
-browse            -hangul_input      +packages          +textprop
++builtin_terms    +iconv             +path_extra        +timers
+byte_offset       +insert_expand     +perl              +title
+channel           +ipv6              +persistent_undo   -toolbar
+cindent           +job               +popupwin          +user_commands
-clientserver      +jumplist          +postscript        +vartabs
+clipboard         +keymap            +printer           +vertsplit
+cmdline_compl     +lambda            +profile           +virtualedit
+cmdline_hist      +langmap           -python            +visual
+cmdline_info      +libcall           +python3           +visualextra
+comments          +linebreak         +quickfix          +viminfo
+conceal           +lispindent        +reltime           +vreplace
+cryptv            +listcmds          +rightleft         +wildignore
+cscope            +localmap          +ruby              +wildmenu
+cursorbind        +lua               +scrollbind        +windows
+cursorshape       +menu              +signs             +writebackup
+dialog_con        +mksession         +smartindent       -X11
+diff              +modify_fname      -sodium            -xfontset
+digraphs          +mouse             -sound             -xim
-dnd               -mouseshape        +spell             -xpm
-ebcdic            +mouse_dec         +startuptime       -xsmp
+emacs_tags        -mouse_gpm         +statusline        -xterm_clipboard
+eval              -mouse_jsbterm     -sun_workshop      -xterm_save
+ex_extra          +mouse_netterm     +syntax            
+extra_search      +mouse_sgr         +tag_binary        
-farsi             -mouse_sysmouse    -tag_old_static    
     系统 vimrc 文件: "$VIM/vimrc"
     用户 vimrc 文件: "$HOME/.vimrc"
 第二用户 vimrc 文件: "~/.vim/vimrc"
      用户 exrc 文件: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
         $VIM 预设值: "/usr/local/share/vim"
编译方式: clang -c -I. -Iproto -DHAVE_CONFIG_H -DMACOS_X -DMACOS_X_DARWIN -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 
链接方式: clang -L. -fstack-protector-strong -L/usr/local/lib -L/usr/local/opt/libyaml/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/readline/lib -L/usr/local/lib -o vim -lm -lncurses -liconv -lintl -framework AppKit -L/usr/local/opt/lua/lib -llua5.4 -mmacosx-version-min=11.3 -fstack-protector-strong -L/usr/local/lib -L/usr/local/Cellar/perl/5.34.0/lib/perl5/5.34.0/darwin-thread-multi-2level/CORE -lperl -L/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/config-3.9-darwin -lpython3.9 -framework CoreFoundation -lruby.3.0 -L/usr/local/Cellar/ruby/3.0.2/lib

**Reply by eirnym** [reply]
Could you please tell us also MacVim's version by running `:version` inside it?

**Comment by dongfangrui** [comment]
MacVim version：Custom version 8.2.2681 （171）
![图片](https://user-images.githubusercontent.com/37399699/125215716-c202f680-e2ee-11eb-9c42-a592ed42ecaa.png)

**Reply by eirnym** [reply]
I prefer version output as you've done for `vim` above by running `mvim --version` where mvim is included in MacVim'a distribution.

**Reply by eirnym** [reply]
I also require an information if you installed `vim` above `MacVim` in homebrew
